### PR TITLE
fix: stop listening for payments when shutting down

### PR DESCRIPTION
**Pull request by m1sterc001guy**
Edge case of gateway's safe stop feature that was pointed out by @elsirion here: https://github.com/fedimint/fedimint/pull/5999/files#r1762742035

There's a race condition when stopping where if the gateway receives a payment request before `cancel_on_shutdown` has stopped the thread, the gateway would start processing the payment and potentially leave the gateway in an un-finished state. This PR just adds a `tokio::select!` to payment processing thread so that it knows to break when the gateway's state is `ShuttingDown`.

### CODE REVIEW

File: gateway/ln-gateway/src/lib.rs
Comment (elsirion) [inline gateway/ln-gateway/src/lib.rs:561]: Not sure if we fully fix raciness, but at least the time where it might happen becomes much shorter. If we start shutdown after the select, but before calling `handle_htlc`, could that still be a problem?

**Review by elsirion — DISMISSED** [review]
Not sure fully fixing it is feasible rn, but this is a step in the right direction (and wayyyy better than the behavior before safe shutdown).

**Comment by elsirion** [comment]
Would like to get a second review though, maybe from @tvolk131 :)

File: gateway/ln-gateway/src/lib.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lib.rs:561]: Currently according to `ILnRpcClient` documentation, not calling `ILnRpcClient::complete_htlc()` for an item sent to the `ILnRpcClient::route_htlcs()` stream is undefined behavior. The underlying behavior for ldk, lnd, and cln is to fail the HTLC after the CLTV expiry time has passed. Definitely not ideal.

@m1sterc001guy can we add a TODO comment in here stating the issue above? This should definitely be addressed down the road.

Also, would it be worth defining the above behavior in `ILnRpcClient`? If so, I can make a PR for that.

Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lib.rs:561]: > If we start shutdown after the select, but before calling handle_htlc, could that still be a problem?

Ah yes, this is still an issue. I think this is fixable by having the section of code after the `select!` protected by a lock, and also the `handle_shutdown_msg` function protected by the same lock. That way, the gateway's state cannot change after we've checked it and call `handle_htlc`

File: gateway/ln-gateway/src/lib.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lib.rs:558]: No. This case doesn't handle the case where the user presses Ctrl-C

Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lib.rs:549]: As discussed offline, I moved this to after the `select!` and after we take the lock, so that we can be sure that `GatewayState` doesn't change before `handle_htlc` is called.

File: gateway/ln-gateway/src/lib.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lib.rs:552]: @elsirion I believe I fixed the remaining issue by holding the `GatewayState` RW Lock. The `handle_shutdown_msg` thread will hold the write lock and this thread will hold the read lock. Here, we now check the state after taking the lock, meaning if we are still in the `Running` state that the state should not change until `handle_htlc` is called (should be atomic now).

Thanks for double checking this!

File: gateway/ln-gateway/src/lib.rs
Comment (tvolk131) [inline gateway/ln-gateway/src/lib.rs:1521]: Can we do something similar here as I mentioned above?

File: gateway/ln-gateway/src/lib.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lib.rs:556]: That does not compile, you cannot deference the `state_guard` because that would move it, and it doesn't implement `Copy`. This is why I am cloning it, so that we can actually read the state.

File: gateway/ln-gateway/src/lib.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lib.rs:1521]: See above comment

File: gateway/ln-gateway/src/lib.rs
Comment (tvolk131) [reply gateway/ln-gateway/src/lib.rs:556]: Did you try it? It runs on my machine. Using `ref lightning_context` and `*state_guard` performs a dereference rather than a clone

Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lib.rs:556]: ah I missed the `ref` in this. Yes this works, I will fix.

File: gateway/ln-gateway/src/lib.rs
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lib.rs:1521]: As discussed on the call, we need to clone here since we need the `LightningContext` in the `ShuttingDown` state.

**Review by tvolk131 — APPROVED** [review]

**Comment by tvolk131** [comment]
Merge conflict 😭

**Comment by m1sterc001guy** [comment]
> Merge conflict

Fixed!

**Review by tvolk131 — APPROVED** [review]

File: gateway/ln-gateway/src/lib.rs
Comment (elsirion) [reply gateway/ln-gateway/src/lib.rs:548]: Looks good to me now!

**Pull request merged**
